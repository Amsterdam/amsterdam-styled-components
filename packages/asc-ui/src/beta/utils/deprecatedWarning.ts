/**
 * A general deprecated console warning avoided by lint
 */

export default (message: string) => {
  // eslint-disable-next-line
  console.warn(message)
}
