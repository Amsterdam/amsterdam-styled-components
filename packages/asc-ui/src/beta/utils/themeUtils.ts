import { css } from 'styled-components'
import type { Theme } from '../types'
import { fromProps } from './fromProps'
import { ascBetaTheme } from '..'

export type ThemeFn<T> = ({ theme }: { theme: Theme.ThemeInterface }) => T

/**
 * Curry function to provide the theme as first parameter, followed up with other parameters
 * @param cb
 *
 * @example
 * const myThemeHelperFunction = withTheme((theme, ...otherParams) => {
 *   ...
 * })
 *
 * css`
 *  ${myThemeHelperFunction(param1, param2)}
 * `
 */
export const withTheme =
  <T extends any[], U = any>(
    cb: (theme: Theme.ThemeInterface, ...params: T) => U,
  ) =>
  (...params: T) =>
  ({ theme }: { theme: Theme.ThemeInterface }): U =>
    cb(theme, ...params)

type ThemeColorParameters = [Theme.ColorType?, string?, string?]

/**
 * A shortcut to the `fromProps` that will get a value out of the props.theme object
 */
export const getValueFromTheme = withTheme<[string, ((value: any) => void)?]>(
  (theme, identifier, callback?: (value: any) => void) =>
    fromProps(identifier, callback)(theme),
)

export const themeColor = withTheme<ThemeColorParameters, string>(
  (theme, colorType, colorSubtype = 'main', override) => {
    if (override) {
      return override
    }

    if (colorType) {
      return getValueFromTheme(`colors.${[colorType]}.${[colorSubtype]}`)({
        theme,
      })
    }

    return getValueFromTheme('colors.tint.level1')({ theme })
  },
)

type BreakpointsType = [Theme.TypeBreakpoint, keyof Theme.BreakpointsInterface]

export const breakpoint = withTheme<BreakpointsType>((theme, type, variant) => {
  const breakpointFunc: Theme.GetBreakpointFunc = getValueFromTheme(
    `breakpoints.${[variant]}`,
  )({
    theme,
  })
  return breakpointFunc && breakpointFunc(type)
})

/**
 * When this style is applied on an element it will be hidden but still readable by screen readers.
 *
 * For example:
 * ```ts
 * const OnlyVisibleForScreenReaders = styled.button`${srOnlyStyle}`
 * ```
 */
export const srOnlyStyle = css`
  border-width: 0;
  clip: rect(0, 0, 0, 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
`

/**
 * Fills the elements in an SVG with a color, useful for styling icons.
 *
 * For example, using a theme color:
 * ```jsx
 * const PrimaryCarIcon = styled(CarIcon)`
 *   ${svgFill(themeColor('primary'))}
 * `
 * ```
 *
 * Or by setting the color directly using any color valid in CSS, such as hexadecimal, rgba() and hsla().
 *
 * ```jsx
 * const RedCarIcon = styled(CarIcon)`
 *   ${svgFill('#FF0000')}
 * `
 * ```
 */
export const svgFill = withTheme((theme, color: string | ThemeFn<string>) => {
  const fill = typeof color === 'function' ? color({ theme }) : color

  return css`
    & svg {
      circle,
      rect,
      polygon,
      path {
        fill: ${fill};
      }
    }
  `
})

export const calculateFluidStyle = (
  minSizePx: number,
  maxSizePx: number,
  minScreenWidthPx = 320,
  maxScreenWidthPx = 1920,
) => {
  const defaultBaseSize = 16
  const minSize = minSizePx / defaultBaseSize
  const maxSize = maxSizePx / defaultBaseSize
  const minScreenWidth = minScreenWidthPx / defaultBaseSize
  const maxScreenWidth = maxScreenWidthPx / defaultBaseSize

  return `clamp(
    ${minSize}rem,
    ${minSize}rem + ${maxSize - minSize} * 
    (100vw - ${minScreenWidth}rem) / ${maxScreenWidth - minScreenWidth},
    ${maxSize}rem
  )`
}

export interface FluidTypoStyle {
  small?: boolean
  intro?: boolean
  styleAs?: any
  as?: any
}

export const fluidTypoStyle = css`
  ${({ small, intro, styleAs, as }: FluidTypoStyle) => {
    const tag =
      (Object.keys(ascBetaTheme.typography).includes(styleAs) && styleAs) ||
      (Object.keys(ascBetaTheme.typography).includes(as) && as) ||
      'p'

    const prop =
      (small &&
        Object.keys(ascBetaTheme.typography[tag]).includes('small') &&
        'small') ||
      (intro &&
        Object.keys(ascBetaTheme.typography[tag]).includes('intro') &&
        'intro') ||
      'default'

    return css`
      font-size: ${calculateFluidStyle(
        ascBetaTheme.typography[tag][prop].minFontSize,
        ascBetaTheme.typography[tag][prop].maxFontSize,
      )};
      line-height: ${calculateFluidStyle(
        ascBetaTheme.typography[tag][prop].minLineHeight,
        ascBetaTheme.typography[tag][prop].maxLineHeight,
      )};
    `
  }}
`

export interface TypoStyle {
  gutterBottom?: {
    small: number
    large: number
  }
  strong?: boolean
  darkBackground?: boolean
}

export const typographyStyle = css<TypoStyle>`
  margin-top: 0;
  margin-bottom: ${({ gutterBottom }) =>
    `${
      typeof gutterBottom?.large === 'number'
        ? gutterBottom.large
        : gutterBottom || 0
    }px`};

  ${({ gutterBottom }) =>
    typeof gutterBottom?.small === 'number' &&
    css`
      @media screen and ${breakpoint('max-width', 'laptop')} {
        margin-bottom: ${gutterBottom.small}px;
      }
    `}

  ${({ strong }) =>
    strong &&
    css`
      font-weight: 700;
    `}

  ${({ darkBackground }) =>
    darkBackground &&
    css`
      color: white;
      ${svgFill('white')}
    `}
`

type ThemeSpacingParameters = [
  Theme.Spacing,
  Theme.Spacing?,
  Theme.Spacing?,
  Theme.Spacing?,
]

/**
 * Retrieve an amount of pixels by passing a factor that will multiply it by the theme's spacing unit
 *
 * @example If the theme's spacing unit is 4px:
 * css`
 *  padding: ${themeSpacing(1, 2, 1, 4)} // padding: 4px 8px 4px 16px;
 *  margin-bottom: ${themeSpacing(3)} // margin-bottom: 12px;
 * `
 */
export const themeSpacing = withTheme<ThemeSpacingParameters>(
  (theme, ...factors) => {
    const spacing: Theme.Spacing = getValueFromTheme('spacing')({ theme })
    return factors
      .map((factor) => factor && `${factor * spacing}px`)
      .join(' ')
      .trim()
  },
)
