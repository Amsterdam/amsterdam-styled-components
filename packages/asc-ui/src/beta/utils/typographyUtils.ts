import { css } from 'styled-components'
import { svgFill, breakpoint } from '.'

export const calculateFluidStyle = (
  minSizePx: number,
  maxSizePx: number,
  minScreenWidthPx = 320,
  maxScreenWidthPx = 1920,
) => {
  const defaultBaseSize = 16
  const minSize = minSizePx / defaultBaseSize
  const maxSize = maxSizePx / defaultBaseSize
  const minScreenWidth = minScreenWidthPx / defaultBaseSize
  const maxScreenWidth = maxScreenWidthPx / defaultBaseSize

  return `clamp(
    ${minSize}rem,
    ${minSize}rem + ${maxSize - minSize} * 
    (100vw - ${minScreenWidth}rem) / ${maxScreenWidth - minScreenWidth},
    ${maxSize}rem
  )`
}

export const fluidTypoConfig = {
  h1: {
    default: {
      minFontSize: 32,
      maxFontSize: 80,
      minLineHeight: 40,
      maxLineHeight: 88,
    },
    small: {
      minFontSize: 30,
      maxFontSize: 60,
      minLineHeight: 34,
      maxLineHeight: 64,
    },
  },
  h2: {
    default: {
      minFontSize: 28,
      maxFontSize: 56,
      minLineHeight: 32,
      maxLineHeight: 64,
    },
  },
  h3: {
    default: {
      minFontSize: 24,
      maxFontSize: 40,
      minLineHeight: 32,
      maxLineHeight: 48,
    },
  },
  h4: {
    default: {
      minFontSize: 22,
      maxFontSize: 32,
      minLineHeight: 32,
      maxLineHeight: 40,
    },
  },
  h5: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 26,
      maxLineHeight: 32,
    },
  },
  h6: {
    default: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 22,
      maxLineHeight: 24,
    },
  },
  p: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 28,
      maxLineHeight: 40,
    },
    intro: {
      minFontSize: 18,
      maxFontSize: 30,
      minLineHeight: 28,
      maxLineHeight: 40,
    },
    small: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 24,
      maxLineHeight: 26,
    },
  },
  a: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 26,
      maxLineHeight: 32,
    },
  },
  button: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 28,
      maxLineHeight: 40,
    },
    small: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 22,
      maxLineHeight: 24,
    },
  },
  blockquote: {
    default: {
      minFontSize: 24,
      maxFontSize: 40,
      minLineHeight: 32,
      maxLineHeight: 48,
    },
  },
  ul: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 28,
      maxLineHeight: 40,
    },
    small: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 24,
      maxLineHeight: 26,
    },
  },
  ol: {
    default: {
      minFontSize: 18,
      maxFontSize: 24,
      minLineHeight: 28,
      maxLineHeight: 40,
    },
    small: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 24,
      maxLineHeight: 26,
    },
  },
  dl: {
    default: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 24,
      maxLineHeight: 26,
    },
  },
  table: {
    default: {
      minFontSize: 16,
      maxFontSize: 18,
      minLineHeight: 24,
      maxLineHeight: 26,
    },
  },
}

export interface FluidTypoStyle {
  small?: boolean
  intro?: boolean
  styleAs?: any
  as?: any
}

export const fluidTypoStyle = css`
  ${({ small, intro, styleAs, as }: FluidTypoStyle) => {
    const tag =
      (Object.keys(fluidTypoConfig).includes(styleAs) && styleAs) ||
      (Object.keys(fluidTypoConfig).includes(as) && as) ||
      'p'

    const prop =
      (small &&
        Object.keys(fluidTypoConfig[tag]).includes('small') &&
        'small') ||
      (intro &&
        Object.keys(fluidTypoConfig[tag]).includes('intro') &&
        'intro') ||
      'default'

    return css`
      font-size: ${calculateFluidStyle(
        fluidTypoConfig[tag][prop].minFontSize,
        fluidTypoConfig[tag][prop].maxFontSize,
      )};
      line-height: ${calculateFluidStyle(
        fluidTypoConfig[tag][prop].minLineHeight,
        fluidTypoConfig[tag][prop].maxLineHeight,
      )};
    `
  }}
`

export interface TypoStyle {
  gutterBottom?: {
    small: number
    large: number
  }
  strong?: boolean
  darkBackground?: boolean
}

export const typographyStyle = css<TypoStyle>`
  margin-top: 0;
  margin-bottom: ${({ gutterBottom }) =>
    `${
      typeof gutterBottom?.large === 'number'
        ? gutterBottom.large
        : gutterBottom || 0
    }px`};

  ${({ gutterBottom }) =>
    typeof gutterBottom?.small === 'number' &&
    css`
      @media screen and ${breakpoint('max-width', 'laptop')} {
        margin-bottom: ${gutterBottom.small}px;
      }
    `}

  ${({ strong }) =>
    strong &&
    css`
      font-weight: 700;
    `}

  ${({ darkBackground }) =>
    darkBackground &&
    css`
      color: white;
      ${svgFill('white')}
    `}
`
