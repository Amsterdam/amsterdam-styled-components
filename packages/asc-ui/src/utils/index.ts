export { default as fromTheme, valueFromObject } from './fromTheme'
export {
  color,
  getColorCode,
  getTypographyFromTheme,
  focusStyleOutline,
  srOnlyStyle,
  breakpoint,
  svgFill,
  showHide,
  ShowHideTypes,
  customCss,
  CustomCssPropsType,
  CustomCssPropsInterface,
} from './themeUtils'
