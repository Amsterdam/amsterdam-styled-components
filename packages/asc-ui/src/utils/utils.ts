import { AbstractNode, IconDefinition } from '@datapunt/asc-icons/lib/types'
import * as React from 'react'

export function log(message: string) {
  if (!(process && process.env && process.env.NODE_ENV === 'production')) {
    console.error(`[icons-react]: ${message}.`)
  }
}

export function isIconDefinition(target: any): target is IconDefinition {
  return (
    typeof target === 'object' &&
    typeof target.name === 'string' &&
    typeof target.icon === 'object'
  )
}

export function normalizeAttrs(attrs: Attrs = {}): Attrs {
  return Object.keys(attrs).reduce((acc: Attrs, key) => {
    const val = attrs[key]
    switch (key) {
      case 'class':
        acc.className = val
        delete acc.class
        break
      default:
        acc[key] = val
    }
    return acc
  }, {})
}

export interface Attrs {
  [key: string]: string
}

export class MiniMap<V> {
  get size() {
    return Object.keys(this.collection).length
  }

  private collection: { [key: string]: V } = {}

  clear(): void {
    this.collection = {}
  }

  delete(key: string): boolean {
    return delete this.collection[key]
  }

  get(key: string): V | undefined {
    return this.collection[key]
  }

  has(key: string): boolean {
    return Boolean(this.collection[key])
  }

  set(key: string, value: V): this {
    this.collection[key] = value
    return this
  }
}

export function generate(
  node: AbstractNode,
  key: string,
  rootProps?: { [key: string]: any } | false,
): any {
  if (!rootProps) {
    return React.createElement(
      node.tag,
      { key, ...normalizeAttrs(node.attrs) },
      (node.children || []).map((child, index) =>
        generate(child, `${key}-${node.tag}-${index}`),
      ),
    )
  }
  return React.createElement(
    node.tag,
    {
      key,
      ...normalizeAttrs(node.attrs),
      ...rootProps,
    },
    (node.children || []).map((child, index) =>
      generate(child, `${key}-${node.tag}-${index}`),
    ),
  )
}
