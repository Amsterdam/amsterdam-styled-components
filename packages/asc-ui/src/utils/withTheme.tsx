import * as React from 'react'
import { shallow, render, mount } from 'enzyme'
import { ThemeProvider } from '../index'

export const renderWithTheme = (component: JSX.Element) =>
  render(<ThemeProvider>{component}</ThemeProvider>)

export const shallowWithTheme = (component: JSX.Element) =>
  shallow(<ThemeProvider>{component}</ThemeProvider>)

export const mountWithTheme = (component: JSX.Element) =>
  mount(<ThemeProvider>{component}</ThemeProvider>)
