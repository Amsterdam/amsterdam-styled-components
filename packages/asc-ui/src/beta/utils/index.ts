import deprecatedWarning from './deprecatedWarning'

export {
  themeColor,
  srOnlyStyle,
  breakpoint,
  svgFill,
  fluidTypoStyle,
  typographyStyle,
  showAboveBackDrop,
  getValueFromTheme,
  themeSpacing,
} from './themeUtils'
export { valueFromObject } from './valueFromObject'
export { fromProps } from './fromProps'

export { deprecatedWarning }
