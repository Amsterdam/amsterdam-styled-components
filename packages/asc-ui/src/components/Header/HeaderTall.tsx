import * as React from 'react'
import { getBreakpointFromTheme } from '../../styles/utils'
import { HeaderStyleProps } from '../../styles/components/HeaderStyle'
import HeaderSearch from './HeaderSearch'
import HeaderMenu from './HeaderMenu'
import HeaderTallTitle from './HeaderTallTitle'
import styled from '../../styled-components'

type Props = HeaderStyleProps & {
  title: string
  homeLink: string
  search: React.ReactElement
  menu: React.ReactElement
}

const HeaderTallStyleWrapper = styled.div`
  background-color: #fff;
  margin: 0 auto;
  max-width: 1800px;
  height: 108px;
  border-bottom: 50px solid #767676;

  @media screen and ${({ theme }) =>
      getBreakpointFromTheme(theme, 'max-width', 'laptop')} {
    display: none;
  }
`

const HeaderTallStyle = styled.div`
  display: flex;
  flex-wrap: wrap;
  min-height: 158px;
  background-color: #fff;
  margin: 0 auto;
  max-width: 1600px;

  & > :first-child {
    height: 108px;
    width: 100%;
  }

  & > :nth-child(2) {
    height: 50px;
    flex-grow: 1;
    background-color: #767676;
  }

  & > :nth-child(3) {
    height: 50px;
    min-width: 100px;
    background-color: #767676;
  }
`

const HeaderTall: React.FC<Props> = ({
  search,
  menu,
  title,
  homeLink,
  ...otherProps
}) => (
  <HeaderTallStyleWrapper>
    <HeaderTallStyle {...otherProps}>
      <HeaderTallTitle title={title} homeLink={homeLink} />
      <HeaderSearch>{search}</HeaderSearch>
      <HeaderMenu>{menu}</HeaderMenu>
    </HeaderTallStyle>
  </HeaderTallStyleWrapper>
)

export default HeaderTall
