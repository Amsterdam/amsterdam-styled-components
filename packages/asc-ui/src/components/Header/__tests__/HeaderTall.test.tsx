import * as React from 'react'
import HeaderTall from '../HeaderTall'
import { renderWithTheme } from '../../../utils/withTheme'
import 'jest-styled-components'

describe('Header tall', () => {
  it('should render the Header tall', () => {
    const component = renderWithTheme(
      <HeaderTall
        title="Data en informatie"
        homeLink="http://data.amsterdam.nl"
        search={<div> Search </div>}
        menu={<div> M </div>}
      />,
    )

    expect(component).toMatchSnapshot()
  })
})
