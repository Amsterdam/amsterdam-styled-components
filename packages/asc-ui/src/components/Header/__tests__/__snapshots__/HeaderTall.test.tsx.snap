// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header tall should render the Header tall 1`] = `
.c0 {
  background-color: #ffffff;
  margin: 0 auto;
  max-width: 1800px;
  height: 108px;
  border-bottom: 50px solid #F5F5F5;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  min-height: 158px;
  background-color: #ffffff;
  margin: 0 auto;
  max-width: 1600px;
}

.c1 > :first-child {
  height: 108px;
  width: 100%;
}

.c1 > :nth-child(2) {
  height: 50px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  background-color: #F5F5F5;
}

.c1 > :nth-child(3) {
  height: 50px;
  min-width: 100px;
  background-color: #F5F5F5;
}

.c4 {
  overflow: hidden;
}

.c2 {
  color: #000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2 > h1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  margin: 0;
  font-family: 'AvenirNextLTW01-Demi';
  font-weight: normal;
}

.c2 > h1 > div > span {
  display: inline-block;
}

.c2 > h1 > div > span {
  margin-top: 2px;
  line-height: 68px;
  font-size: 30px;
}

.c3 {
  display: inline-block;
  min-height: 108px;
  padding: 20px 40px 20px 15px;
}

.c3 > svg {
  height: 68px;
  width: 100px;
}

@media screen and (max-width:1024px) {
  .c0 {
    display: none;
  }
}

@media screen and (max-width:768px) {
  .c4 {
    width: 50px;
  }
}

@media screen and (min-width:768.02px) {
  .c4 {
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
  }
}

@media screen and (max-width:1024px) {
  .c5 {
    width: 50px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <a
      class="c2"
      href="http://data.amsterdam.nl"
    />
    <h1>
      <a
        class="c2"
        href="http://data.amsterdam.nl"
      />
      <a
        class="c3"
        href="http://data.amsterdam.nl"
      >
        SVG_MOCK
      </a>
      <div>
        <span>
          Data en informatie
        </span>
      </div>
    </h1>
    <div
      class="c4"
    >
      <div>
         Search 
      </div>
    </div>
    <div
      class="c5"
    >
      <div>
         M 
      </div>
    </div>
  </div>
</div>
`;
